File: java/multicast/Receiver.java
import java.io.IOException;
import java.net.DatagramPacket;
import java.net.MulticastSocket;
import java.net.SocketTimeoutException;
import java.net.InetAddress;

/**
 * Receiver represents a UDP packet receiver with multicast capabilities: several receivers can join
 * the multicast group and receive packets intended for the multicast address.
 * @author Thomas Samy Dafir
 */
public class Receiver {
    
    private final static int MAXSIZE = 512;
    private final static String GROUP_IP = "230.0.0.1";

    /**
     * main handles wrong numbers of cmd-arguments and calls the receive-method
     * @param args cmd-arguments:
     * 	args[0] port number
     * 	args[1] timeout for the receiving socket
     * @throws IOException
     */
	public static void main(String[] args) throws IOException {
		if(args.length != 2){
			System.out.println("Usage: java Receiver <port number> <receiving timeout in ms>");
		}else{
			receive(args);
		}
	}
	
	/**
	 * creates a multicast group as well as group address and a socket. joins the socket to
	 * the multicast group and receives UDP packets. The content of the packets is saved in
	 * the buffer byte-array. Also counts the amount of packets received.
	 * @param args cmd-arguments passed from main
	 * @throws IOException exception is forwarded to be handled by calling method
	 */
	private static void receive(String[] args) throws IOException{
		byte[] buffer = new byte[MAXSIZE];
		MulticastSocket socket = new MulticastSocket(Integer.parseInt(args[0]));
        InetAddress group = InetAddress.getByName(GROUP_IP);
        socket.joinGroup(group);
		socket.setSoTimeout(Integer.parseInt(args[1]));
		DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
		int packetCount = 0;
		try{
			while(true){
				socket.receive(packet);
				//print(packet);
				packetCount++;
			}
		}catch(SocketTimeoutException e){
			System.out.println("Receiving of packets terminated");
			System.out.println(packetCount + " packets received");
		}
        socket.leaveGroup(group);
		socket.close();
	}
	
	
	/**
	 * Prints the content of a UDP packet
	 * @param packet packet to be printed
	 */
	@SuppressWarnings("unused")
	private static void print(DatagramPacket packet){
		String message = new String(packet.getData(), 0, packet.getLength());
		System.out.println(message);
	}
}

File: c/multicast/receiver.c
/**
 * Receiver represents a UDP packet receiver with multicast capabilities: several receivers can join
 * the multicast group and receive packets intended for the multicast address.
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <time.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include "receiver.h"

#define MAXSIZE 512
#define GROUP_IP "230.0.0.1"

int main(int argc, char *argv[]){

    //requied local var declaration
    int sock, sockaddr_len, one, t_out, rec_cnt;
    struct sockaddr_in receiver, sender;
    struct ip_mreq multi;
    struct timeval timeout;
    char buffer[MAXSIZE];    
    bzero(&receiver, sizeof(receiver));
    bzero(&multi, sizeof(receiver));

    //create udp ipv4 socket
    sock = socket(AF_INET,SOCK_DGRAM,0);
    if(sock < 0){
        handle_error("socket could not be created");
    }

    //Allow multiple receivers to use same socket
    setsockopt(sock,SOL_SOCKET,SO_REUSEADDR,&one, sizeof(one));
    
    //set address options, like protocol, socket,...
    receiver.sin_family = AF_INET;
    receiver.sin_addr.s_addr = htonl(INADDR_ANY);
    receiver.sin_port = htons(atoi(argv[1]));
    
    //configure timeout for socket
    timeout.tv_sec = atoi(argv[2]);
    timeout.tv_usec = 0;
    if(setsockopt(sock,SOL_SOCKET,SO_RCVTIMEO,&timeout,sizeof(timeout)) < 0){
        handle_error("could not set socket receive-timeout");
    }

    //bind socket to created address
    bind(sock,(struct sockaddr*)&receiver,sizeof(receiver));

    //set a multicast address and join socket to the group for that address
    multi.imr_multiaddr.s_addr = inet_addr(GROUP_IP);
    multi.imr_interface.s_addr = htonl(INADDR_ANY);
    setsockopt(sock,IPPROTO_IP,IP_ADD_MEMBERSHIP,(const void *)&multi, sizeof(struct ip_mreq));

    //calculate sizeof sockaddr struct
    sockaddr_len = sizeof(struct sockaddr_in);

    //enter while loop and receive packets from the socket and save them into the buffer
    rec_cnt = 0;
    while(recvfrom(sock,buffer,MAXSIZE,0,(struct sockaddr*)&sender,&sockaddr_len) >= 0){
        rec_cnt++;
    }
    printf("%d packets received \n", rec_cnt);
    close(sock);
}

void handle_error(char *message){
    puts(message);
    exit(1);
}
